src/router/BaseRouter.php

```diff
@@ -27,21 +27,19 @@ abstract class BaseRouter<+TResponder> {
       $data = Dict\map($data, $value ==> \urldecode($value));
       return tuple($responder, new ImmMap($data));
     } catch (NotFoundException $e) {
-      foreach (HttpMethod::getValues() as $next) {
-        if ($next === $method) {
-          continue;
-        }
-        try {
-          list($responder, $data) = $resolver->resolve($next, $path);
-          if ($method === HttpMethod::HEAD && $next === HttpMethod::GET) {
-            $data = Dict\map($data, $value ==> \urldecode($value));
-            return tuple($responder, new ImmMap($data));
-          }
-          throw new MethodNotAllowedException();
-        } catch (NotFoundException $_) {
-          continue;
+      $allowed = $this->getAllowedMethods($path);
+      if (0 !== C\count($allowed)) {
+        if (
+          $method === HttpMethod::HEAD && C\contains($allowed, HttpMethod::GET)
```

Comment:
how about:

- remove the outer if and assignments
- replace this condition with `$method === HttpMethod::HEAD && $this->getAllowedMethods($path) === keyset[HttpMethod::GET]` ?

Reply:
> remove the outer if 

we need the if statment to determine whether or not to throw `MethodNotAllowedException`, if there isn't a method that is allowed, we throw `NotFoundException`

> and assignments

we pass the allowed methods to exception later on if we are not re-routing `HEAD` <-> `GET`

> replace this condition with $method === HttpMethod::HEAD && $this->getAllowedMethods($path) === keyset[HttpMethod::GET] 

the same path can accept multiple http methods, we dispatch `GET` handler for `GET`, `HEAD`, `GET, POST`, `GET, PUT, PATCH` ... etc

Reply:
sorry, yes, the assignemnt and if are needed - but please refactor to not increase the nesting - invert the condition and exit early (by throwing)

> the same path can accept multiple http methods, we dispatch GET handler for GET, HEAD, GET, POST, GET, PUT, PATCH ... etc

This is what the old code did, but it is unsafe as we were discussing last week. The only safe HEAD -> GET fallback is if GET is the only registered responder for that path

Reply:
done.